File: tests/suid_sudo_test_passenv.rb
#!/usr/bin/ruby
if __FILE__ != $0
  p "run directly."
  exit(1)
end
require File.absolute_path("../suid_sudo", File.dirname(__FILE__)).untaint
# require_relative claims insecure operation

include SUID_SUDO::INCLUDE
p suid_emulate(sudo_wrap:true, use_shebang:false, pass_env:['TESTVAR'], showcmd_opts:true)

cmd = ARGV[0] || "0"

def _ids
  p [[Process::Sys::getuid, Process::Sys::geteuid],
     [Process::Sys::getgid, Process::Sys::getegid],
     Process.groups]
end
def _print_ids
  p _ids
end
def _print_ids_env
  p _ids
  ENV.each {|k,v| print "#{k}=#{v}\n" }
end

case cmd
when 's'
  p SUID_SUDO.const_get(:SUID_STATUS_)::_status
when 'p'
  print ("\nbe USER\n")
  temporarily_as_user()
  _print_ids_env

  print ("\nbe ROOT\n")
  temporarily_as_root()
  _print_ids_env

  print ("\nbe REAL_ROOT\n")
  temporarily_as_real_root()
  _print_ids_env

  print ("\nbe USER\n")
  temporarily_as_user()
  _print_ids_env

  print ("\ndrop to USER\n")
  drop_privileges_forever()
  _print_ids_env

  begin
    print ("\nbe ROOT\n")
    temporarily_as_root()
    _print_ids_env
    print ("SHOULD FAILED!\n")
  rescue SUIDPrivilegesSettingError => e
    print("...good to be failed: #{e}\n")
  end
when 'pb'
  print "\ntemporary_in_user\n"
  temporarily_as_user {
    _print_ids_env
  }
  print "\nreturned\n"
  _print_ids_env
  print "temporary_in_root\n"
  temporarily_as_root {
    _print_ids_env
  }
  print "\nreturned\n"
  _print_ids_env
end

File: tests/suid_sudo_test_passenv.py
#!/usr/bin/python3 -I
if __name__ != '__main__': raise RuntimeError("Should not be imported.")

import sys
import os.path
sys.path.insert(0, os.path.dirname(sys.argv[0]) + "/..")
import suid_sudo
from suid_sudo import *
_SuidStatus = suid_sudo._SuidStatus

import os
def print_ids():
    print (os.getresuid(), os.getresgid(), os.getgroups())

suid_emulate(sudo_wrap=True, inherit_flags=True, realroot_ok=True, nonsudo_ok=True,
             pass_env = ["TESTVAR"], showcmd_opts=True)

cmd = sys.argv[1] if len(sys.argv) >= 2 else "0"

if cmd == "p":
    print ("\nbe USER ")
    temporarily_as_user()
    print_ids()
    for k in os.environ:
        print ("%s=%s" % (k, os.environ[k]))

    print ("\nbe ROOT")
    temporarily_as_root()
    print_ids()
    for k in os.environ:
        print ("%s=%s" % (k, os.environ[k]))

    print ("\nbe REAL_ROOT")
    temporarily_as_real_root()
    print_ids()
    for k in os.environ:
        print ("%s=%s" % (k, os.environ[k]))

    print ("\nbe USER")
    temporarily_as_user()
    print_ids()
    for k in os.environ:
        print ("%s=%s" % (k, os.environ[k]))

    print ("\ndrop to USER")
    drop_privileges_forever()
    print_ids()
    for k in os.environ:
        print ("%s=%s" % (k, os.environ[k]))

    try:
        print ("\nbe ROOT")
        temporarily_as_root()
        print_ids()
        print ("SHOULD FAILED!")
        raise
    except (SUIDPrivilegesSettingError, OSError) as e:
        print("...good to be failed: %r" % e)

elif cmd == "s":
    print (_SuidStatus._status)

elif cmd == "pw":
    print ("\nbe USER ")
    with temporarily_as_user():
        print_ids()
        for k in os.environ:
            print ("%s=%s" % (k, os.environ[k]))

    print ("\nreturned ")
    print_ids()
    for k in os.environ:
        print ("%s=%s" % (k, os.environ[k]))

    print ("\nbe ROOT")
    with temporarily_as_root():
        print_ids()
        for k in os.environ:
            print ("%s=%s" % (k, os.environ[k]))

    print ("\nbe REAL_ROOT")
    with temporarily_as_real_root():
        print_ids()
        for k in os.environ:
            print ("%s=%s" % (k, os.environ[k]))

    print ("\nbe USER")
    with temporarily_as_user():
        print_ids()
        for k in os.environ:
            print ("%s=%s" % (k, os.environ[k]))

    print ("\ndrop to USER")
    try:
        with drop_privileges_forever():
            print_ids()
            for k in os.environ:
                print ("%s=%s" % (k, os.environ[k]))
        print ("SHOULD FAILED!")
        raise
    except (SUIDPrivilegesSettingError, OSError) as e:
        print("...good to be failed: %r" % e)

else:
    print("unknown test command")
